flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/internal/BackchannelImpl.java

```diff
@@ -0,0 +1,198 @@
+     * concurrent queue when the writer and committer are not chained.
+     */
+    private Thread writerThread;
+    /**
+     * The committer thread that consumes messages from the writer. Used to ensure that we switch to
+     * a concurrent queue when the writer and committer are not chained.
+     */
+    private Thread committerThread;
+    /**
+     * The messages to be sent from the writer to the committer. By default, it's a thread-safe
+     * implementation but it can be replaced with an unsynchronized queue when the writer and
+     * committer are chained (common case).
+     */
+    private Deque<T> messages = new ConcurrentLinkedDeque<>();
+    /** Flag to indicate that the backchannel has been established. */
+    private boolean established;
+
+    BackchannelImpl(Tuple2<Integer, String> key) {
+        this.key = key;
+    }
+
+    @Override
+    public void send(T message) {
+        messages.add(message);
+    }
+
+    @Override
+    public void consume(Consumer<T> consumer) {
+        ensureEstablished();
+        T message;
+        while ((message = pollUnsafe()) != null) {
+            consumer.accept(message);
+        }
+    }
+
+    @Nullable
+    private T pollUnsafe() {
+        return messages.poll();
+    }
+
+    @VisibleForTesting
+    @Nullable
+    public T poll() {
+        ensureEstablished();
+        return pollUnsafe();
+    }
+
+    @VisibleForTesting
+    @Override
+    public BackchannelImpl<T> establish() {
+        return create(key.f0, key.f1, writerThread == null);
+    }
+
+    /**
+     * Create a backchannel for the given subtask id and transactional id prefix. The backchannel is
+     * created if it does not exist, otherwise, the existing backchannel is returned.
+     *
+     * <p>If this method is called twice with the same subtask id and transactional id prefix but
+     * different forWriter values, the backchannel will be established and can be used for
+     * communication.
+     *
+     * <p>If this method is called twice with the same subtask id and transactional id prefix and
+     * the same forWriter value, an error will be logged instead. There is a high chance that the
+     * backchannel is misused by using non-uniques transactionalIds.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> BackchannelImpl<T> create(
+            int subtaskId, String transactionalIdPrefix, boolean forWriter) {
+        Tuple2<Integer, String> key = new Tuple2<>(subtaskId, transactionalIdPrefix);
+        BackchannelImpl<T> backchannel;
+        boolean duplicate;
+        synchronized (BACKCHANNELS) {
+            backchannel =
+                    (BackchannelImpl<T>)
+                            BACKCHANNELS.computeIfAbsent(key, k -> new BackchannelImpl<>(key));
+            duplicate = backchannel.isDuplicate(forWriter);
+            backchannel.register(forWriter);
+        }
+        if (duplicate) {
+            LOG.error(
```

Comment:
Is it intentionally that the logging happens outside the critical section? It makes reading the code a lot hard.

Can we do
```java
if (backchannel.isDuplicate(forWriter)) {
  LOG.error(...)
}
```

Reply:
Yeah I can do that. My mind set is sometimes still in the time of the runtime team where each critical section was bad. So sometimes, I optimize too much given that these are once during recovery or once during checkpoint calls.